File: Python/hotel.py
from random import randrange
from datetime import datetime

class Hotel:
    def __init__(self):
        self.rooms = []
        self.reservations = []
        self.availability_calendar = [[True] * 365 for _ in self.rooms]
        self.initialize_availability_calendar()

    def add_room(self, room):
        self.rooms.append(room)

    def calculate_total_cost(self, room, check_in_date, check_out_date):
        price_per_night = room.get_price_per_night()
        number_of_nights = self.calculate_number_of_nights(check_in_date, check_out_date)
        return price_per_night * number_of_nights

    def calculate_number_of_nights(self, check_in_date, check_out_date):
        check_in_date = datetime.strptime(check_in_date, "%Y-%m-%d")
        check_out_date = datetime.strptime(check_out_date, "%Y-%m-%d")
        difference = check_out_date - check_in_date
        number_of_nights = difference.days
        return max(number_of_nights, 1)

    def generate_confirmation_number(self):
        current_time = datetime.now().strftime("%Y%m%d%H%M%S")
        random_number = randrange(10000)
        confirmation_number = f"CONF{current_time}{random_number}"
        return confirmation_number

    def is_room_available(self, room_number, check_in_date, check_out_date):
        for reservation in self.reservations:
            reserved_room = reservation.get_room()
            reserved_check_in = reservation.get_check_in_date()
            reserved_check_out = reservation.get_check_out_date()

            if reserved_room.get_room_number() == room_number:
                if (
                    check_in_date >= reserved_check_in
                    and check_in_date < reserved_check_out
                ) or (check_out_date > reserved_check_in and check_out_date <= reserved_check_out):
                    return False
        return True

    def list_reservations(self):
        print("List of Reservations:")
        if not self.reservations:
            print("No reservations have been made.")
        else:
            for reservation in self.reservations:
                self.print_reservation_confirmation(reservation)
        print()

    def make_reservation(self, customer, room_number, check_in_date, check_out_date):
        if not self.is_room_available(room_number, check_in_date, check_out_date):
            print("Room is not available for the specified dates.")
            return

        room = None
        for r in self.rooms:
            if r.get_room_number() == room_number:
                room = r
                break

        if room is None:
            print("Room not found.")
            return

        reservation = Reservation(customer, room, check_in_date, check_out_date)
        self.reservations.append(reservation)
        self.update_availability_calendar(reservation)
        print("Reservation made successfully!")

    def initialize_availability_calendar(self):
        self.availability_calendar = [[True] * 365 for _ in self.rooms]

    def update_availability_calendar(self, reservation):
        room = reservation.get_room()
        check_in_date = reservation.get_check_in_date()
        check_out_date = reservation.get_check_out_date()

        room_index = -1
        for i, r in enumerate(self.rooms):
            if r.get_room_number() == room.get_room_number():
                room_index = i
                break

        if room_index == -1:
            print("Error: Room not found.")
            return

        check_in_date = datetime.strptime(check_in_date, "%Y-%m-%d")
        check_out_date = datetime.strptime(check_out_date, "%Y-%m-%d")

        check_in_day = check_in_date.timetuple().tm_yday
        check_out_day = check_out_date.timetuple().tm_yday

        while len(self.availability_calendar[room_index]) < 365:
            self.availability_calendar[room_index].append(True)

        for day in range(check_in_day - 1, check_out_day):
            self.availability_calendar[room_index][day] = False


class Room:
    def __init__(self, room_number, room_type, price_per_night, max_occupancy, amenities):
        self.room_number = room_number
        self.is_occupied = False
        self.room_type = room_type
        self.price_per_night = price_per_night
        self.max_occupancy = max_occupancy
        self.amenities = amenities

    def get_room_type(self):
        return self.room_type

    def get_room_number(self):
        return self.room_number

    def get_price_per_night(self):
        return self.price_per_night

    def get_max_occupancy(self):
        return self.max_occupancy

    def get_amenities(self):
        return self.amenities

class Reservation:
    def __init__(self, customer, room, check_in_date, check_out_date):
        self.customer = customer
        self.room = room
        self.check_in_date = check_in_date
        self.check_out_date = check_out_date

    def get_customer(self):
        return self.customer

    def get_room(self):
        return self.room

    def get_check_in_date(self):
        return self.check_in_date

    def get_check_out_date(self):
        return self.check_out_date

if __name__ == "__main__":
    hotel = Hotel()

File: C++/Hotel.cpp
#include "Hotel.h"
#include <iostream>
#include <iomanip>
#include <ctime>
#include <sstream>
#include "Reservation.h"

void Hotel::printReservationConfirmation(const Reservation &reservation)
{
    const Customer &customer = reservation.getCustomer();
    const Room &room = reservation.getRoom();
    const std::string &checkInDate = reservation.getCheckInDate();
    const std::string &checkOutDate = reservation.getCheckOutDate();

    double totalCost = calculateTotalCost(room, checkInDate, checkOutDate);

    std::cout << "Reservation Confirmation" << std::endl;
    std::cout << "Customer Name: " << customer.getName() << std::endl;
    std::cout << "Customer Email: " << customer.getEmail() << std::endl;
    std::cout << "Room Number: " << room.getRoomNumber() << std::endl;
    std::cout << "Check-In Date: " << checkInDate << std::endl;
    std::cout << "Check-Out Date: " << checkOutDate << std::endl;
    std::cout << "Total Cost: $" << totalCost << std::endl;
    std::cout << "Confirmation Number: " << generateConfirmationNumber() << std::endl;
    std::cout << std::endl;
}

void Hotel::updateAvailabilityCalendar(const Reservation &reservation)
{
    const Room &room = reservation.getRoom();
    const std::string &checkInDate = reservation.getCheckInDate();
    const std::string &checkOutDate = reservation.getCheckOutDate();

    int roomIndex = -1;
    for (size_t i = 0; i < rooms.size(); ++i)
    {
        if (rooms[i].getRoomNumber() == room.getRoomNumber())
        {
            roomIndex = static_cast<int>(i);
            break;
        }
    }

    if (roomIndex == -1)
    {
        std::cerr << "Error: Room not found." << std::endl;
        return;
    }

    struct tm tm_in = {};
    struct tm tm_out = {};
    strptime(checkInDate.c_str(), "%Y-%m-%d", &tm_in);
    strptime(checkOutDate.c_str(), "%Y-%m-%d", &tm_out);
    int checkInDay = tm_in.tm_yday;
    int checkOutDay = tm_out.tm_yday;

    for (int day = 0; day < 365; ++day)
    {
        if (day >= checkInDay && day <= checkOutDay)
        {
            availabilityCalendar[roomIndex][day] = false; // Room not available on this day
        }
    }
}

void Hotel::initializeAvailabilityCalendar()
{
    availabilityCalendar.resize(rooms.size(), std::vector<bool>(365, true));
}

Hotel::Hotel()
{
    initializeAvailabilityCalendar();
}

void Hotel::addRoom(const Room &room)
{
    rooms.push_back(room);
}

double Hotel::calculateTotalCost(const Room &room, const std::string &checkInDate, const std::string &checkOutDate)
{
    double pricePerNight = room.getPricePerNight();
    int numberOfNights = calculateNumberOfNights(checkInDate, checkOutDate);
    return pricePerNight * numberOfNights;
}

int Hotel::calculateNumberOfNights(const std::string &checkInDate, const std::string &checkOutDate)
{
    std::tm checkInTm = {};
    std::tm checkOutTm = {};
    std::istringstream(checkInDate) >> std::get_time(&checkInTm, "%Y-%m-%d");
    std::istringstream(checkOutDate) >> std::get_time(&checkOutTm, "%Y-%m-%d");

    std::time_t checkInTime = std::mktime(&checkInTm);
    std::time_t checkOutTime = std::mktime(&checkOutTm);

    std::time_t difference = checkOutTime - checkInTime;

    int numberOfNights = static_cast<int>(difference / (60 * 60 * 24));

    // User must stay at least 1 night
    return (numberOfNights > 0) ? numberOfNights : 1;
}

std::string Hotel::generateConfirmationNumber()
{

    std::time_t currentTime = std::time(nullptr);

    int randomNumber = rand() % 10000;
    std::stringstream confirmationNumber;
    confirmationNumber << "CONF" << std::put_time(std::localtime(&currentTime), "%Y%m%d%H%M%S") << randomNumber;

    return confirmationNumber.str();
}

bool Hotel::isRoomAvailable(int roomNumber, const std::string &checkInDate, const std::string &checkOutDate)
{

    for (const Reservation &reservation : reservations)
    {
        const Room &reservedRoom = reservation.getRoom();
        const std::string &reservedCheckIn = reservation.getCheckInDate();
        const std::string &reservedCheckOut = reservation.getCheckOutDate();

        if (reservedRoom.getRoomNumber() == roomNumber)
        {
            if ((checkInDate >= reservedCheckIn && checkInDate < reservedCheckOut) ||
                (checkOutDate > reservedCheckIn && checkOutDate <= reservedCheckOut))
            {
                return false;
            }
        }
    }

    return true;
}

void Hotel::listReservations()
{
    std::cout << "List of Reservations:" << std::endl;

    if (reservations.empty())
    {
        std::cout << "No reservations have been made." << std::endl;
    }
    else
    {
        for (const Reservation &reservation : reservations)
        {
            printReservationConfirmation(reservation);
        }
    }
    std::cout << std::endl;
}

void Hotel::makeReservation(Customer customer, int roomNumber, const std::string &checkInDate, const std::string &checkOutDate)
{
    // Check room availability here:
    if (!isRoomAvailable(roomNumber, checkInDate, checkOutDate))
    {
        std::cout << "Room is not available for the specified dates." << std::endl;
        return;
    }

    Room *room = nullptr;
    for (Room &r : rooms)
    {
        if (r.getRoomNumber() == roomNumber)
        {
            room = &r;
            break;
        }
    }

    if (room == nullptr)
    {
        std::cout << "Room not found." << std::endl;
        return;
    }

    Reservation reservation(customer, *room, checkInDate, checkOutDate);

    reservations.push_back(reservation);

    updateAvailabilityCalendar(reservation);

    std::cout << "Reservation made successfully!" << std::endl;
}
